include/fast_ber/ber_types/Real.hpp

```diff
@@ -50,4 +63,502 @@ class Real
+ * @param input BER encoded real binary value to check
+ * @return true on success, false on failure
+ */
+inline bool check_real_binary(absl::Span<const uint8_t> input) noexcept
+{
+    /* Minimal binary construction is: | control | exponent | number |  => length is 3 bytes minimum
+     * Check that base bits are not set to the reserved value, see X690 Clause 8.5.7.4
+     * */
+    if (input.length() > 3  && (input[0] & 0x30) != 0x30)
+    {
+        /* X690 Clause 8.5.7.4 */
+        size_t exponent_len = (input[0] & 0x03) + 1;
+        size_t exponent_offset = 1;
+
+        /* X690 Clause 8.5.7.4.d: we should get something like this:
+         * | control | exponent_len | exponent_1 | ... | exponent_N | number |
+         * */
+        if (exponent_len == 4)
+        {
+            exponent_len = static_cast<size_t>(input[1]);
+            exponent_offset = 2;
+        }
+
+        /* Check that exponent length is valid or does not overflow input */
+        if (exponent_len == 0 || (exponent_len + exponent_offset) >= input.length())
+        {
+            return false;
+        }
+
+        /* Check that the exponent does not start with 9 identical bits */
+        if (exponent_len > 1)
+        {
+            uint8_t first_byte = input[exponent_offset];
+            uint8_t second_byte = input[exponent_offset + 1];
+            if ((first_byte == 0xFF && (second_byte & 0x80) != 0) ||
+                (first_byte == 0x00 && (second_byte & 0x80) == 0))
+            {
+                /* The exponent begins with 9 identical bits */
+                return false;
+            }
+        }
+
+        /* Check that the mantissa is not zero */
+        for (size_t i = exponent_len + exponent_offset ; i < input.length() ; ++i)
+        {
+            if (input[i] != 0) return true;
+        }
+    }
+
+    return false;
+}
+
+
+/**
+ * Decodes a real value into a double
+ * @param input BER encoded real value
+ * @param output decoded special value
+ * @return true on success, false on failure
+ */
+inline bool decode_real(absl::Span<const uint8_t> input, double& output) noexcept
+{
+    bool res = false;
+    std::array<uint8_t, sizeof(double)> buffer;
+
+    /* X690 Clause 8.5.2: it is the value Plus Zero */
+    if (input.length() == 0)
+    {
+        output = 0.0;
+        res = true;
+    }
+    /* X690 Clause 8.5.9: it is a Special Real Value */
+    else if (input[0] & 0x40)
+    {
+        res = decode_real_special(input, output);
+    }
+    /* X690 Clause 8.5.7: it is a Binary value */
+    else if (input[0] & 0x80)
+    {
+        res = decode_real_binary(input, output);
+    }
+    /* X690 Clause 8.5.8: it is a Decimal value */
+    else if ((input[0] & 0xC0) == 0)
+    {
+        res = decode_real_decimal(input, output);
+    }
+
+    /* Swap bytes */
+    if (res)
+    {
+        std::memcpy(buffer.data(), &output, sizeof(double));
+        std::reverse(buffer.begin(), buffer.end());
```

Comment:
Why reverse?

Reply:
Network byte order to host byte order. 
I suggest detecting little/big endian on the machine which runs the ber codec ?

Reply:
Fine.

This should only apply to decode_real_binary, real_decimal and real_specail will already be host order.

It may be more wise to use [frexp](http://www.cplusplus.com/reference/cmath/frexp/) / [ldexp](http://www.cplusplus.com/reference/cmath/ldexp/) to avoid the bit twiddling altogether